lib/src/services/bedrock_service.dart

```diff
@@ -3,31 +3,67 @@ import 'package:dio/dio.dart';
 import 'package:flutter_secure_storage/flutter_secure_storage.dart';
 import 'package:jwt_decoder/jwt_decoder.dart';
 
+import 'dio_logger.dart';
+
+final Dio dio = Dio();
+
 class BedrockNetworkInterceptor extends Interceptor {
   final IAuthTokenStorage storage = AuthStorage(const FlutterSecureStorage());
 
+  static String TAG = "services_base_api";
+
   @override
   void onRequest(
     RequestOptions options,
     RequestInterceptorHandler handler,
   ) async {
-    String token = await storage.readAuthToken() ?? '';
+    String? token = await storage.readAuthToken(); //get token
     Map<String, dynamic> headers = {
       'Accept': 'application/json',
       'content-type': 'application/json',
     };
 
-    try {
+    if (token != null) {
       if (!JwtDecoder.isExpired(token)) {
         await storage.storeAuthToken(token);
         options.headers.addAll({...headers, 'Authorization': 'Bearer $token'});
+      } else {
+        dio.interceptors.requestLock
+            .lock(); // lock the request to wait for new token to be given
+        dio.interceptors.responseLock
+            .lock(); // lock the request to wait for new token to be given
+
+        //TODO fetch the refresh token from the storage
+        //TODO ask for a new token to the backend using the stored refresh token
+        //TODO set the new token + refresh token again in the storage and put the token in the headers
+
+        dio.interceptors.requestLock
+            .unlock(); // unlock the request when token is given
+        dio.interceptors.responseLock
+            .unlock(); // unlock the request when token is given
       }
-    } catch (e) {
-      options.headers.addAll(headers);
+    } else {
+      options.headers
+          .addAll(headers); //authentication phase doesn't have a token yet
     }
 
     super.onRequest(options, handler);
   }
+
+  @override
+  void onResponse(Response response, ResponseInterceptorHandler handler) {
+    DioLogger.onSuccess(TAG, response);
+    super.onResponse(response, handler);
+  }
+
+  @override
+  void onError(DioError err, ErrorInterceptorHandler handler) {
+    DioLogger.onError(TAG, err);
```

Comment:
1. Does `DioLogger` output its values to the console only or in a file somewhere?
2. In a future PR, we're planning to add environments (dev, staging, production). I imagine logging will probably need to be disabled when running in staging/production environment and only work in dev mode. Is there an easy way to toggle on/off logging?

Reply:
Yes it's totally possible. I can maybe add environments in this PR and make this logic in the `DioLogger`